andela/Ah-backend-xmen
# #162725265 Users should be able to receive notifications

#### What does this PR do?
This PR handles in-app notifications and email notifications.
Notifications give users to opportunity to know what they may have missed which may be important to them.


#### Description of Task to be completed?
There may be 2 types of notifications: email notifications and in-app notifications.

Email notifications are sent as email messages to the email address the user submitted during registration.
Users should be able to opt out (and in) of this type of notification.

In-app notifications show up in the app. This may be visible on a notification page or some other view, as the team may choose.

#### How should this be manually tested?

In-app notifications can be created, read, and marked as read. 

To Test this follow the steps below:
 - Fetch and check into this branch with `git fetch origin ft-notifications-162725265 && git checkout ft-notifications-162725265`
 - setup and activate your virtual environment
 - make migrations with `python manage.py migrate`
 - run the server with `python manage.py runserver`
 - Test the following endpoints with Postman

This PR so far addresses has the following endpoints
`GET /api/notifications/all` 
This fetches all notifications of an authenticated user

`GET /api/notifications/unread`
This fetched unread notifications

`PUT /api/notifications/all`
This marks all notifications as read

##### notification permisions:

`GET /api/notifications/perms<token>` 
 User is opted out of notificaion type encoded in token


`PUT /api/notifications/perms/<notification_type>`
User opts into a specific notification type such as `PROFILE_FOLLOW` and `RECEIVE_NOTIFICATION_EMAILS`

#### What are the relevant pivotal tracker stories?
[162725265](https://www.pivotaltracker.com/story/show/162725265)

#### FUNCTIONALITY TESTING INSTRUCTIONS

1. Login to your author's haven account (use a valid email for this)
2. Fetch all notifications at `GET /api/notifications/all` 
3. The result should be an empty notifications array
4. Create an article and save its slug in your environment variable
5. Log in as a different User and Leave a comment on the created article
6. Log back in as the author and 'GET /api/notifications/unread'
7. You should see your in-app notifications
8. To mark them as read, `PUT /api/notifications/all`
9. Log into the author's email client such as [gmail](mail.google.com) to see the email notification.
10. copy the notification opt-out URL from the email message and paste it in the browser
11. You should see a successful opt-out message
12. To opt back into the notification do `PUT /api/notifications/perms/<notification_type>` as an
authenticated user.


##### POSTMAN
[![Run in Postman](https://run.pstmn.io/button.svg)](https://app.getpostman.com/run-collection/e76f0d7329a979325f26)
